backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -50,30 +54,96 @@ public Match findMatch(Participant newParticipant) {
-      long currEndTimeAvailable = currParticipant.getEndTimeAvailable();
-      long earliestEndTimeAvailable = Math.min(newEndTimeAvailable, currEndTimeAvailable);
-      boolean compatibleTime =
-          (currentTimeMillis + TimeUnit.MINUTES.toMillis(duration + PADDING_MINUTES))
-              < earliestEndTimeAvailable;
-
-      if (compatibleTime) {
-        // Create and return match
-        return new Match(
-            newParticipant.getUsername(),
-            currParticipant.getUsername(),
-            duration,
-            currentTimeMillis);
+      long secondEndTimeAvailable = currParticipant.getEndTimeAvailable();
+      if (!isCompatibleTime(duration, firstEndTimeAvailable, secondEndTimeAvailable)) {
+        System.out.println("not compatible time");
+        continue;
+      }
+
+      // Check match preference compatibility and get combined preference if compatible
+      MatchPreference secondMatchPreference = currParticipant.getMatchPreference();
+      MatchPreference combinedMatchPreference =
+          getCombinedMatchPreference(firstMatchPreference, secondMatchPreference);
+      if (combinedMatchPreference == null) {
+        System.out.println("not compatible match pref");
+        continue;
+      }
+
+      // Check if combined match preference is satisfied depending on number of same fields
+      if (!isCombinedMatchPreferenceSatisfied(
+          combinedMatchPreference, newParticipant, currParticipant)) {
+        continue;
       }
+
+      // Found a match
+      System.out.println("found a match");
+      return new Match(
+          newParticipant.getUsername(), currParticipant.getUsername(), duration, clock.millis());
     }
     // No inital match found
     return null;
   }
+
+  /** Return true if compatible endTimeAvailable considering duration and padding */
+  private boolean isCompatibleTime(
+      int duration, long firstEndTimeAvailable, long secondEndTimeAvailable) {
+    long earliestEndTimeAvailable = Math.min(firstEndTimeAvailable, secondEndTimeAvailable);
+    return (clock.millis() + TimeUnit.MINUTES.toMillis(duration + PADDING_MINUTES))
+        < earliestEndTimeAvailable;
+  }
+
+  /**
+   * @return combined MatchPreference similar = both are similar OR one is similar and one is any;
+   *     any = both are any; different = both are different = one is different and one is any. or
+   *     null if match preferences are not compatible
+   */
+  private MatchPreference getCombinedMatchPreference(
+      MatchPreference firstMatchPreference, MatchPreference secondMatchPreference) {
+    if ((firstMatchPreference == secondMatchPreference)
+        || (firstMatchPreference == MatchPreference.ANY)
+        || (secondMatchPreference == MatchPreference.ANY)) {
+      return (firstMatchPreference == MatchPreference.ANY)
+          ? secondMatchPreference
+          : firstMatchPreference;
+    } else {
+      return null;
+    }
```

Comment:
This is a bit difficult to parse. WDYT about flipping the condition

```suggestion
  private MatchPreference getCombinedMatchPreference(
      MatchPreference firstMatchPreference, MatchPreference secondMatchPreference) {
    if ((firstMatchPreference != secondMatchPreference)
        && (firstMatchPreference != MatchPreference.ANY)
        && (secondMatchPreference != MatchPreference.ANY)) {
       	// One user is SIMILAR, other is DIFFERENT. Incompatible.
       	return null;
    }
    return (firstMatchPreference == MatchPreference.ANY)
          ? secondMatchPreference
          : firstMatchPreference;

```

Reply:
See the comment I tagged you in above in Tiancong's comment.

Reply:
Thanks I like Tiancong's suggestion better.